File: connie_giannopoulos/week_01/EX5_LAB_MTA/MTA.js
// The Subway: Lines N, L & 6 and their stations.
var subwayLines = {
  'N' : ['Times Square', '34th', '28th', '23rd', 'Union Square','8th'],
  'L' : ['8th', '6th', 'Union Square', '3rd', '1st'],
  '6' : ['Grand Central', '33rd', '28th', '23rd', 'Union Square','Astor Place']
};



// Prompts.

var userInput = function(planTrip) {

  var options = {  
    startLine: prompt('Please pick your starting line: N, L or 6.'),
    startStop: prompt('Which station would you like to start your journey at?'),
    endLine: prompt('Please pick your end line: N, L or 6.'),
    endStop: prompt('Which station do you need to get off at?')
  }

  return options;

}

var tripParams = userInput();



// Plan trip function.

var planTrip = function(startLine, startStop, endLine, endStop) {
    
    var startStopIndex;
    var endStopIndex;
    var route;
    
    // If the startLine and endLine are the same.
    if(startLine === endLine) {
        
        // Create references to the index of startStop and endStop.
        startStopIndex = subwayLines[startLine].indexOf(startStop);
        endStopIndex = subwayLines[endLine].indexOf(endStop);
        // console.log(startStopIndex);
        // console.log(endStopIndex);
        route = subwayLines[startLine].splice(startStopIndex, endStopIndex + 1);
        
        console.log('You need to travel on the ' + startLine + ' line: ' + route.join(', ') + '\n');
        
    // If startLine and endLine are different.
    } else {
        
        var intersection = true;
        // Create references to the index of startStop and endStop.
        // Create a slice for startStopIndex / endStopIndex to Union Square
       // var startLineUnionIndex
        // create the route variable;
        
        startStopIndex = subwayLines[startLine].indexOf(startStop);
        var startStopUnionIndex = subwayLines[startLine].indexOf('Union Square');
        
        // If indexes are in wrong order,
        // Swap their values.
        // With Array.prototype.slice(), the first index need must be smaller,
        // than the second index.
        if(startStopIndex > startStopUnionIndex) {
            startStopIndexHolder = startStopIndex;
            startStopIndex = startStopUnionIndex;
            startStopUnionIndex = startStopIndexHolder;
        }
        
        endStopIndex = subwayLines[endLine].indexOf(endStop);
        var endStopUnionIndex = subwayLines[endLine].indexOf('Union Square');
         
        // If indexes are in wrong order,
        // Swap their values.
        // With Array.prototype.slice(), the first index need must be smaller,
        // than the second index.
        if(endStopIndex > endStopUnionIndex) {
            endStopIndexHolder = endStopIndex;
            endStopIndex = endStopUnionIndex;
            endStopUnionIndex = endStopIndexHolder;
        }
        //console.log(endStopIndex);
        //console.log(endStopUnionIndex);
        
        var startRoute = subwayLines[startLine].slice(startStopIndex, startStopUnionIndex + 1);
        var endRoute = subwayLines[endLine].slice(endStopIndex, endStopUnionIndex + 1);
        
        route = startRoute.concat(endRoute);
        
        console.log('You need to travel from the ' + startLine + ' line: ' + startRoute.join(', ') + '.' + '\n');
    
        console.log('You need to change at Union Square.' + '\n');

    
        console.log('Your journey continues through the following stops: ' + endRoute.join(', ') + '.' + '\n');
    }
    
    console.log('There are ' + route.length + ' stops in total.' + '\n')
    
    //console.log('Thank you for using this amazing subway!');
    var chooChoo = [
      '___________   _______________________________________^__',
      ' ___   ___ |||  ___   ___   ___    ___ ___  |   __  ,----\\',
      '|   | |   |||| |   | |   | |   |  |   |   | |  |  | |_____\\',
      '|___| |___|||| |___| |___| |___|  | O | O | |  |  |        \\',
      '           ||| Choo Choo!         |___|___| |  |__|         )',
      '___________|||______________________________|______________/',
      '           |||          All aboard the Connie train..   /--------',
      '-----------\'\'\'---------------------------------------'
      ];

    chooChoo.forEach(function(line) {
        console.log(line);
    });
};

// If you need to change at `Union Square` over two lines.
planTrip(tripParams.startLine, tripParams.startStop, tripParams.endLine, tripParams.endStop);


// If you travel on the same lane.
//planTrip('L', '8th', 'L', '3rd');

// If you travel on the same lane.
//planTrip('N', 'Union Square', '6', 'Astor Place');




























 

File: drew_showalter/week_03/tuesday/mta.rb
require 'pry'
# MTA Lab
# Objectives:
# Apply your knowledge of Ruby to solve a real world problem.
# Get really good at array manipulation.
# Activity
# Create a program that models a simple subway system.
# The program takes the line and stop that a user is getting on at and the line and stop that user is getting off at and prints the journey and the total number of stops for the trip in the console:
# planTrip('N', 'Times Square', '6', '33rd'); // This is only a suggested function name and signature.
# // console.log() shows output similar to this:
# // "Your must travel through the following stops on the N line: 34th, 28th, 23rd, Union Square."
# // "Change at Union Square."
# // "Your journey continues through the following stops: 23rd, 28th, 33rd."
# // "7 stops in total."
# There are 3 subway lines:
n_line = { :stops => "Times Square, 34th, 28th, 23rd, Union Square, 8th".split(', '), :name => 'The N Line' }
l_line = { :stops => "8th, 6th, Union Square, 3rd, 1st".split(', '), :name => 'The L Line' }
six_line = { :stops => "Grand Central, 33rd, 28th, 23rd, Union Square, Astor Place".split(', '), :name => 'The 6 Line' }
all_lines = [n_line, l_line, six_line]
# All 3 subway lines intersect at Union Square, but there are no other intersection points. (For example, this means the 28th stop on the N line is different than the 28th street stop on the 6 line, so you'll have to differentiate this when you name your stops in the arrays.)
# Tell the user the number of stops AND the stops IN ORDER that they will pass through or change at.
# Hints:

def plan_trip(line1, start, line2, stop)
  if line1 == line2
    puts "You travel along these stops: "
    p travel(line1[:stops], find_index(line1, start), find_index(line1, stop))
  else
    puts "You travel along these stops:"
    p travel(line1[:stops], find_index(line1, start), find_index(line1, "Union Square"))
    puts "changing at Union Square you continue along:"
    p ( travel(line2[:stops], find_index(line2, "Union Square"), find_index(line2, stop)) )[1..-1]
  end
end

# def find_line(stop)
#   container = []
#   all_lines.each do |hash|
#     if hash[:stops].include?
#       container << hash[:name]
#     end
#   end
#   if container.length > 1
#     p "Which line are you stopping at?: "
#     name = gets.chomp
#   end
# end

def find_index(line, stop)
  line[:stops].index(stop)
end

def travel(line_stops, index1, index2)
  if index1 < index2
    line_stops[index1..index2]
  else
    line_stops[index2..index1].reverse
  end
end
#travel(n_line[:stops], 4, 0)

plan_trip(l_line, '8th', l_line, '1st')
plan_trip(l_line, '8th', n_line,'Times Square')

binding.pry
# Work out how you would do it on paper first! Then start to explain that process in Ruby.
# Get the program to work for a single line before trying to tackle multiple lines.
# Don't worry about prompting the user for input. Hard code some values to get it working. You can use prompt() later to make it more interactive.
# Consider diagramming the lines by sketching out the subway lines and their stops and intersection.
# The key to the lab is finding the index positions of each stop. (hint: indexOf())
# Make sure the stops that are the same for different lines have different names (i.e. 23rd on the N and on the 6 need to be differentiated)